Incoming flow errors

I'm in the process of finalizing the added pattern matching support for the flow analysis. I used to think this wouldn't lead to new errors, but I discovered there's a missing piece of the analysis along the way.

I tried the (currently in-progress) analysis and it so far only found one new problem:

```
Errors for edu:umn:cs:melt:ableC:abstractsyntax:substitution
 [./edu.umn.cs.melt.ableC/abstractsyntax/substitution/Attribute.sv]
Attribute.sv:60:2: warning: Synthesized equation 'substituted' exceeds flow type with dependencies on env, returnType
```

This is the relevant production:

```
aspect production appliedAttrib
top::Attrib ::= n::AttribName  e::Exprs
{
  local substitutions::Substitutions = top.substitutions;
  substitutions.nameIn =
    case n, e of
      attribName(n), consExpr(stringLiteral(s), nilExpr()) ->
        ...
    | _, _ -> ""
    end;
  top.substituted =
    ... substitutions.refIdSub ...;
}
```

Notice we're pattern matching on `e`, which means we must be able to evaluate the forwarding equation of any expressions, which presumably requires `env, returnType`.

I believe this was missed before because we're pattern matching on `Exprs` outermost which possibly has a forwarding flow type that's empty, so the outer pattern match was OK. Then matching on the inner child erroneously believed it had a proper `Decorated Expr` and assumed the ref set. But that's not a safe assumption without checking.

So with it now "looking through" pattern matching, it sees that the inner child only has `env` if the outer has `env`, so it accurately emits the dependency, which causes the flow type for `substituted` to go awry.

Anyway, I think this implementation isn't done yet, but already it's found one new flow error. Let's hope it doesn't find any more. Just a heads-up.

More progress/fixes. Current status:

```
Errors for edu:umn:cs:melt:ableC:abstractsyntax
 [./edu.umn.cs.melt.ableC/abstractsyntax/AttributeAnimation.sv]
AttributeAnimation.sv:5:2: warning: Equation has transitive dependency on child attr's inherited attribute for env but this equation appears to be missing.
AttributeAnimation.sv:5:2: warning: Equation has transitive dependency on child attr's inherited attribute for returnType but this equation appears to be missing.
AttributeAnimation.sv:21:2: warning: Equation has transitive dependency on child attr's inherited attribute for env but this equation appears to be missing.
AttributeAnimation.sv:21:2: warning: Equation has transitive dependency on child attr's inherited attribute for returnType but this equation appears to be missing.

Errors for edu:umn:cs:melt:ableC:abstractsyntax:substitution
 [./edu.umn.cs.melt.ableC/abstractsyntax/substitution/Attribute.sv]
Attribute.sv:60:2: warning: Synthesized equation substituted exceeds flow type with dependencies on env, returnType
```

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -44,14 +44,16 @@ node {
     /* stages are pretty much just labels about what's going on */
     stage ("Build") {
       /* Clean Silver-generated files from previous builds in this workspace */
-      sh "mkdir -p generated"
       sh "rm -rf generated/* || true"
 
       checkout scm
 
+      /* Checkout sometimes(?) cleans out the workspace, so create this afterwards */
+      sh "mkdir -p generated"
+
       /* env.PATH is the master's path, not the executor's */
       withEnv(["PATH=${params.SILVER_BASE}/support/bin/:${env.PATH}"]) {
-        sh "./build -G ${WORKSPACE}/generated --warn-all --warn-error"
+        sh "./build -G generated --warn-all --warn-error"
       }
     }
 
@@ -125,19 +127,23 @@ node {
         sh "./runTests"
       }
     }
-	} catch (e) {
-		currentBuild.result = 'FAILURE'
-		throw e
-	} finally {
+
+    /* If we've gotten all this way with a successful build, don't take up disk space */
+    sh "rm -rf generated/* || true"
+
+  } catch (e) {
+    currentBuild.result = 'FAILURE'
+    throw e
+  } finally {
     def previousResult = currentBuild.previousBuild?.result
 
-		if (currentBuild.result == 'FAILURE') {
-			notifyBuild(currentBuild.result)
-		} else if (currentBuild.result == null &&
-        previousResult && previousResult == 'FAILURE') {
-			notifyBuild('BACK_TO_NORMAL')
+    if (currentBuild.result == 'FAILURE') {
+      notifyBuild(currentBuild.result)
+    } else if (currentBuild.result == null &&
+      previousResult && previousResult == 'FAILURE') {
+      notifyBuild('BACK_TO_NORMAL')
     }
-	}
+  }
 }
 
 /* Slack / email notification
@@ -145,15 +151,15 @@ node {
  * https://bitbucket.org/snippets/fahl-design/koxKe */
 def notifyBuild(String buildStatus = 'STARTED') {
   // build status of null means successful
-  buildStatus =  buildStatus ?: 'SUCCESSFUL'
+  buildStatus = buildStatus ?: 'SUCCESSFUL'
 
   // Default values
   def colorName = 'RED'
   def colorCode = '#FF0000'
   def subject = "${buildStatus}: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]'"
   def summary = "${subject} (${env.BUILD_URL})"
-  def details = """<p>STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]':</p>
-    <p>Check console output at &QUOT;<a href='${env.BUILD_URL}'>${env.JOB_NAME} [${env.BUILD_NUMBER}]</a>&QUOT;</p>"""
+  def details = """<p>${buildStatus}: Job '${env.JOB_BASE_NAME} ${env.BRANCH_NAME} [${env.BUILD_NUMBER}]':</p>
+    <p>Check console output at <a href='${env.BUILD_URL}'>${env.BUILD_URL}</a></p>"""
 
   // Override default values based on build status
   if (buildStatus == 'STARTED') {
@@ -171,10 +177,11 @@ def notifyBuild(String buildStatus = 'STARTED') {
   slackSend (color: colorCode, message: summary)
 
   emailext(
-      subject: subject,
-      body: details,
-			to: 'evw@umn.edu',
-      recipientProviders: [[$class: 'CulpritsRecipientProvider']]
-    )
+    subject: subject,
+    mimeType: 'text/html',
+    body: details,
+    to: 'evw@umn.edu',
+    recipientProviders: [[$class: 'CulpritsRecipientProvider']]
+  )
 }
 

--- a/edu.umn.cs.melt.ableC/abstractsyntax/AttributeAnimation.sv
+++ b/edu.umn.cs.melt.ableC/abstractsyntax/AttributeAnimation.sv
@@ -2,6 +2,9 @@
 function animateAttributeOnType
 Type ::= attr::Attributes  ty::Type
 {
+  -- TODO: These are hacks we should probably improve upon.
+  attr.env = emptyEnv();
+  attr.returnType = nothing();
   return
     case attr of
     | consAttribute(gccAttribute(l), t) -> animateAttribOnType(l, animateAttributeOnType(t, ty))
@@ -18,6 +21,9 @@ Type ::= attr::Attributes  ty::Type
 function animateAttribOnType
 Type ::= attr::Attribs  ty::Type
 {
+  -- TODO: These are hacks we should probably improve upon.
+  attr.env = emptyEnv();
+  attr.returnType = nothing();
   return case attr of
   -- Attribs that specify new types (i.e. not type-compatible with what is wrapped)
   -- __vector_size__(num)

--- a/edu.umn.cs.melt.ableC/abstractsyntax/substitution/Attribute.sv
+++ b/edu.umn.cs.melt.ableC/abstractsyntax/substitution/Attribute.sv
@@ -50,9 +50,14 @@ top::Attrib ::= n::AttribName
 aspect production appliedAttrib
 top::Attrib ::= n::AttribName  e::Exprs
 {
+  -- TODO: These are hacks we should probably improve upon.
+  local treat_e_syntactically :: Exprs = e;
+  treat_e_syntactically.env = emptyEnv();
+  treat_e_syntactically.returnType = nothing();
+  
   local substitutions::Substitutions = top.substitutions;
   substitutions.nameIn =
-    case n, e of
+    case n, treat_e_syntactically of
       attribName(n), consExpr(stringLiteral(s), nilExpr()) ->
         if n.name == "refId" then substring(1, length(s) - 1, s) else ""
     | _, _ -> ""

